File: macos/Sources/FreePlayer/LRC.swift
import Foundation

struct LyricLine: Identifiable, Equatable {
    var id: Int { time.hashValue }
    var time: Double
    var text: String
}

enum LRC {

    // Parse raw LRC text into sorted, deduped lyric lines.
    // Handles [mm:ss.xx]text, [mm:ss]text, multi-timestamp lines, metadata tags.
    static func parse(_ raw: String?) -> [LyricLine] {
        guard let raw, !raw.isEmpty else { return [] }
        var entries: [LyricLine] = []

        let lines = raw.components(separatedBy: .newlines)
        for rawLine in lines {
            let trimmed = rawLine.trimmingCharacters(in: .whitespaces)
            if trimmed.isEmpty { continue }

            var times: [Double] = []
            var scanIndex = trimmed.startIndex
            let pattern = "\\[(\\d{1,3}):(\\d{2})(?:[.:](\\d{1,3}))?\\]"
            while let range = trimmed.range(of: pattern, options: .regularExpression, range: scanIndex..<trimmed.endIndex) {
                let tag = String(trimmed[range])
                scanIndex = range.upperBound
                let parts = tag.dropFirst().dropLast().split(separator: ":", maxSplits: 1)
                guard let m = Int(parts[0]), let s = Int(parts[1].prefix(2)) else { continue }
                var cs = 0.0
                let fracPart = parts[1].dropFirst(2)
                if !fracPart.isEmpty {
                    let c = Int(fracPart.prefix(2)) ?? 0
                    cs = Double(c) / 100.0
                }
                times.append(Double(m * 60) + Double(s) + cs)
            }
            if times.isEmpty { continue }

            guard let lastBracket = trimmed.lastIndex(of: "]") else { continue }
            let textStart = trimmed.index(after: lastBracket)
            let text = sanitize(String(trimmed[textStart...]).trimmingCharacters(in: .whitespaces))
            if text.isEmpty { continue }

            for time in times {
                entries.append(LyricLine(time: time, text: text))
            }
        }

        entries.sort { $0.time < $1.time }
        var deduped: [LyricLine] = []
        for e in entries {
            if let last = deduped.last, last.text == e.text { continue }
            deduped.append(e)
        }
        return deduped
    }

    // Replace control / invisible / non-renderable glyphs with spaces.
    static func sanitize(_ text: String) -> String {
        var result = ""
        for scalar in text.unicodeScalars {
            let cp = scalar.value
            if cp <= 0x08 || cp == 0x0B || cp == 0x0C
                || (cp >= 0x0E && cp <= 0x1F)
                || (cp >= 0x7F && cp <= 0x9F)
                || (cp >= 0x200B && cp <= 0x200F)
                || (cp >= 0x2028 && cp <= 0x202E)
                || (cp >= 0x2060 && cp <= 0x206F)
                || cp == 0xFEFF || cp == 0xFFFD {
                result.append(" ")
            } else {
                result.unicodeScalars.append(scalar)
            }
        }
        return result
    }

    // Read an .lrc file, auto-detecting CJK encodings.
    // Chain mirrors the Kotlin port: UTF-8 → GB18030/GBK → Shift_JIS → Big5 → EUC-KR.
    // Like the web/Kotlin versions, GB18030 can "successfully" decode Shift_JIS
    // bytes into garbage, so when both decode cleanly we prefer Shift_JIS unless
    // its output is full of half-width katakana (a misdecoded GBK/GB18030 file).
    static func read(path: String) -> String? {
        guard let data = FileManager.default.contents(atPath: path) else { return nil }
        if let s = String(data: data, encoding: .utf8) { return s }

        let gb18030 = String.Encoding(rawValue: CFStringConvertEncodingToNSStringEncoding(
            CFStringEncoding(CFStringEncodings.GB_18030_2000.rawValue)))
        let gbk = String.Encoding(rawValue: CFStringConvertEncodingToNSStringEncoding(
            CFStringEncoding(CFStringEncodings.GBK_95.rawValue)))
        let big5 = String.Encoding(rawValue: CFStringConvertEncodingToNSStringEncoding(
            CFStringEncoding(CFStringEncodings.big5.rawValue)))
        let eucKR = String.Encoding(rawValue: CFStringConvertEncodingToNSStringEncoding(
            CFStringEncoding(CFStringEncodings.EUC_KR.rawValue)))

        let gb = String(data: data, encoding: gb18030) ?? String(data: data, encoding: gbk)
        let sjis = String(data: data, encoding: .shiftJIS)

        if let sjis {
            let halfWidthKana = sjis.unicodeScalars.contains { $0.value >= 0xFF61 && $0.value <= 0xFF9F }
            if gb == nil || !halfWidthKana {
                return sjis
            }
        }
        if let gb { return gb }

        if let s = String(data: data, encoding: big5) { return s }
        if let s = String(data: data, encoding: eucKR) { return s }
        return nil
    }
}

File: src/utils/lrc.js
/**
 * Parse raw LRC text into an array of { time, text } entries.
 * Handles: [mm:ss.xx]text, [mm:ss]text, multi-timestamp lines,
 * and metadata tags like [ti:], [ar:], [al:], [length:].
 */
export function parseLRC(raw) {
  if (!raw) return [];

  const lines = raw.split(/\r?\n/);
  const entries = [];

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed) continue;

    // Collect all time tags on this line
    const timeRegex = /\[(\d{1,3}):(\d{2})(?:[.:](\d{1,3}))?\]/g;
    const times = [];
    let match;
    while ((match = timeRegex.exec(trimmed)) !== null) {
      const minutes = parseInt(match[1], 10);
      const seconds = parseInt(match[2], 10);
      const centiseconds = match[3]
        ? parseInt(match[3].padEnd(2, '0').slice(0, 2), 10)
        : 0;
      times.push(minutes * 60 + seconds + centiseconds / 100);
    }

    if (times.length === 0) continue; // metadata line, skip

    // Get the text after the last time tag
    const textStart = trimmed.lastIndexOf(']') + 1;
    const text = sanitizeText(trimmed.slice(textStart).trim());
    if (!text) continue; // skip empty lyric lines

    for (const time of times) {
      entries.push({ time, text });
    }
  }

  // Sort by time
  entries.sort((a, b) => a.time - b.time);

  // Deduplicate adjacent entries with same text
  const deduped = [];
  for (let i = 0; i < entries.length; i++) {
    if (i === 0 || entries[i].text !== entries[i - 1].text) {
      deduped.push(entries[i]);
    }
  }

  return deduped;
}

/**
 * Sanitize lyric text: replace control chars and non-renderable glyphs
 * with spaces so we never show tofu (□) or ? as placeholders.
 */
export function sanitizeText(text) {
  if (!text) return text;
  let result = '';
  for (let i = 0; i < text.length; i++) {
    const cp = text.codePointAt(i);
    // Skip surrogate pair trailing half so we don't double-process
    if (cp > 0xFFFF) i++;
    if (
      cp <= 0x08 ||                           // C0 controls (except \t=0x09, \n=0x0A, \r=0x0D)
      cp === 0x0B || cp === 0x0C ||           // VT, FF
      (cp >= 0x0E && cp <= 0x1F) ||           // rest of C0
      (cp >= 0x7F && cp <= 0x9F) ||           // DEL + C1 controls
      (cp >= 0x200B && cp <= 0x200F) ||       // zero-width space & joiners
      (cp >= 0x2028 && cp <= 0x202E) ||       // line/paragraph sep, bidi controls
      (cp >= 0x2060 && cp <= 0x206F) ||       // word joiner, invisible operators
      cp === 0xFEFF ||                        // BOM / zero-width no-break space
      cp === 0xFFFD                           // replacement character
    ) {
      result += ' ';
    } else {
      result += text[i];
    }
  }
  return result;
}
